app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php

```diff
@@ -32,34 +32,57 @@ public function __construct(ScopeConfigInterface $scopeConfig)
     }
 
     /**
-     * Put all javascript to footer before sending the response.
+     * Moves all JavaScript tags to the end of body if this feature is enabled.
      *
      * @param Http $subject
      * @return void
      */
     public function beforeSendResponse(Http $subject)
     {
-        $content = $subject->getContent();
-        $script = [];
-        if (is_string($content) && strpos($content, '</body') !== false) {
-            if ($this->scopeConfig->isSetFlag(
-                self::XML_PATH_DEV_MOVE_JS_TO_BOTTOM,
-                ScopeInterface::SCOPE_STORE
-            )
-            ) {
-                $pattern = '#<script[^>]*+(?<!text/x-magento-template.)>.*?</script>#is';
-                $content = preg_replace_callback(
-                    $pattern,
-                    function ($matchPart) use (&$script) {
-                        $script[] = $matchPart[0];
-                        return '';
-                    },
-                    $content
-                );
-                $subject->setContent(
-                    str_replace('</body', implode("\n", $script) . "\n</body", $content)
-                );
+        $content = (string)$subject->getContent();
+        $bodyEndTag = '</body';
+        $isEndBodyTagFound = strpos($content, $bodyEndTag) !== false;
+        $shouldMoveJsToBottom = $this->scopeConfig->isSetFlag(
+            self::XML_PATH_DEV_MOVE_JS_TO_BOTTOM,
+            ScopeInterface::SCOPE_STORE
+        );
+
+        if ($isEndBodyTagFound && $shouldMoveJsToBottom) {
+            if ($scripts = $this->extractScriptTags($content)) {
+                $content = str_replace($bodyEndTag, "$scripts\n$bodyEndTag", $content);
+                $subject->setContent($content);
             }
         }
     }
+
+    /**
+     * Extracts and returns script tags found in given content.
+     *
+     * @param string $content
+     */
+    public function extractScriptTags(&$content)
```

Comment:
Is there any reason of making this method public and not declaring return type?
```suggestion
    private function extractScriptTags(&$content): string
```